### PR TITLE
Add type hints and restyle docstrings

### LINKED ISSUE #31: Add type hints
Add type hints to all functions. Simultaneously, remove type hints from function docstrings as this information does not need to be duplicated.

### PR DESCRIPTION
**Pull request by mary-cleaton**
## Description

Add type hints to all functions. In particular, please can you check:

- Type hints are correct (e.g. Spark session is `pyspark.sql.SparkSession` not `spark.sql.SparkSession`).
- Type hints are consistent (e.g. `cutpoints` is always `Dict[str, Union[List[float], None]])`, `input_variables` is always `Dict[str, Any]`).
- Type hinting is removed from docstrings in both Args and Returns sections.

I have also standardised the docstrings, where I noticed that they were non-standard, and updated the docstring styling (mostly by reducing the indent to two spaces from four, to reduce file length). Standardising the docstrings included standardising subheading order, which as per our chosen style should be:

- Args
- Dependencies
- Provisos
- Returns
- Attribution

## Type of change

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

**Comment by mary-cleaton** [comment]
Wait til #63 merged, then rebase and add type hints to run script. Then, this can be published and ready for review.

**Comment by mary-cleaton** [comment]
Rebased and did some updates, but #63 still not in `develop`, so leaving this as draft until that happens. Then, will rebase again and do any minor tweaks needed before publishing this MR.

**Comment by mary-cleaton** [comment]
Forgot that the branch in #65 was a child of #63 and contained its updates. As #65 has been merged, #63 was closed without merge.

The upshot is that this branch (i.e. #73) can be published, as it has already been rebased against the code it was waiting for (it's just that that code was added to `develop` via #65 rather than #63).

File: scalelink/indicator_matrix/indicator_matrix.py [hunk 1]
```diff
@@ -344,62 +353,66 @@ def compare_deltas(df, linkage_vars, delta_col_prefix):
...
-        string2 (Pyspark Data Frame col: str)
-            string to be compared to string1.
+      string1:
+        First string column to be compared.
+      string2:
+        Second string column to be compared.
 
     Returns:
-        Normalized levenshtein distance between 0 and 1.
+      levenshtein_normalized_distance:
```
Comment (danasebo) [inline scalelink/indicator_matrix/indicator_matrix.py]: change to "normalised" here as well

File: tests/match_scores/test_match_scores.py [hunk 2]
```diff
@@ -50,11 +52,11 @@ def test_assign_match_score(spark):
...
-    spark,
-    assign_weights_input_df,
-    assign_weights_input_x_star,
-    assign_weights_output_df,
-):
+    spark: pyspark.sql.SparkSession,
+    assign_weights_input_df: pyspark.sql.DataFrame,
+    assign_weights_input_x_star: Dict[str, float],
+    assign_weights_output_df: pyspark.sql.DataFrame,
+) -> None:
```
Comment (danasebo) [inline tests/match_scores/test_match_scores.py:59]:
```
    Dependencies:
        chispa as ch
```
on L64-65 needs indent to be updated

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 3]
```diff
@@ -33,7 +33,7 @@
 from scalelink.matrix_a_star import matrix_a_star as ma
 
 
-def test_calculate_b():
+def test_calculate_b() -> None:
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py:36]: `Dependencies:` sections in this script need their indent to be updated

File: scalelink/utils/utils.py [hunk 4]
```diff
@@ -85,49 +88,54 @@ def define_binary_agreement_vars(cutpoints):
...
+      df1_path:
+        The filepath for the first dataframe.
+      df2_path:
+        The filepath for the second dataframe.
+      spark:
+        The Spark session being used.
 
     Dependencies
-        pyspark
+      pyspark
```
Comment (danasebo) [inline scalelink/utils/utils.py]: shouldn't this be `pyspark.sql.SparkSession`, as for the following function?

File: scalelink/main.py [hunk 5]
```diff
@@ -12,7 +13,7 @@
 from scalelink.utils import utils as ut
 
 
-def run_scalelink(config_path="scalelink/configs.ini"):
+def run_scalelink(config_path: str = "scalelink/configs.ini") -> pyspark.sql.DataFrame:
     """
```
Comment (danasebo) [inline scalelink/main.py:17]: why is `shutil` as a dependency of this function? I can't see it used anywhere.

Comment (danasebo) [inline scalelink/main.py:17]: also, indentation of this docstring need an update for this function

File: scalelink/indicator_matrix/indicator_matrix.py [hunk 6]
```diff
@@ -431,27 +444,27 @@ def get_deltas(df_cartesian_join, input_variables):
...
-            single character, col_bigrams will be Null.
+      df:
+        The dataframe containing the two columns that the Sorensen-Dice
+        coefficent will be calculated for.
+      col:
+        The name of the first string column to be compared.
 
     Dependencies:
-        The feature transformer NGram from pyspark.ml.feature.
+      from pyspark.ml.feature import NGram
```
Comment (danasebo) [inline scalelink/indicator_matrix/indicator_matrix.py]: could this be written as `pyspark.ml.feature.NGram`?

File: scalelink/indicator_matrix/indicator_matrix.py [hunk 7]
```diff
@@ -217,45 +224,47 @@ def calculate_deltas(df, cutpoints, agreement_col_suffix, string_similarity_suff
...
+        The number of decimal places new_col should have. Must be a positive
+        number.
 
-    Returns:
-        df_compared (Spark DataFrame):
-            A dataframe consisting of df with an additional float column called
-            new_col which contains the Sorensen-Dice coefficients for the
-            comparison of col1 and col2.
+    Dependencies:
+      from pyspark.ml.feature import NGram
```
Comment (danasebo) [inline scalelink/indicator_matrix/indicator_matrix.py]: could this be written as `pyspark.ml.feature.NGram`?

File: scalelink/indicator_matrix/indicator_matrix.py [hunk 8]
```diff
@@ -217,45 +224,47 @@ def calculate_deltas(df, cutpoints, agreement_col_suffix, string_similarity_suff
...
+      from scalelink.indicator_matrix.indicator_matrix import make_bigrams.
```
Comment (danasebo) [inline scalelink/indicator_matrix/indicator_matrix.py]: could this be written as `scalelink.indicator_matrix.indicator_matrix.make_bigrams`?

File: scalelink/matrix_a_star/matrix_a_star.py [hunk 9]
```diff
@@ -339,29 +347,29 @@ def label_x_star(x_star, cutpoints):
...
     Args:
-        df_njklm (Pandas DataFrame):
-            A dataframe containing the delta comparison column names as column
-            names and a single row consisting of the Njklm values (i.e. the sums
-            of these delta comparison columns).
-        K (int):
-            The total number of agreement states across all linkage variables.
-        p (int):
-            The total number of linkage variables.
+      df_njklm:
```
Comment (danasebo) [inline scalelink/matrix_a_star/matrix_a_star.py]: docstring says it is `df_njklm`, however function uses `Njklm` - I would suggest to use `df_njklm` everywhere

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 10]
```diff
@@ -144,7 +144,7 @@ def test_label_x_star():
     assert test_output == expected_output
 
 
-def test_make_matrix_a():
+def test_make_matrix_a() -> None:
     """
     Tests that make_matrix_a() gives the correct output when provided with
     appropriate inputs.
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py:150]:
```
    Dependencies:
        numpy as np
        pandas as pd
```
on L152-154 need indentation update

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 11]
```diff
@@ -209,7 +209,7 @@ def test_make_matrix_a():
     np.testing.assert_array_equal(test_output, expected_output)
 
 
-def test_make_matrix_a_star():
+def test_make_matrix_a_star() -> None:
     """
     Tests that make_matrix_a_star() gives the correct output when provided with
     appropriate inputs.
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py:215]:
```
    Dependencies:
        numpy as np
```

on L217 need indentation update

File: tests/matrix_a_star/test_matrix_a_star.py [hunk 12]
```diff
@@ -318,7 +318,7 @@ def test_scale_x_star():
     assert test_output == expected_output
 
 
-def test_solve_for_x_star():
+def test_solve_for_x_star() -> None:
     """
     Tests that solve_for_x_star() gives the correct output when provided with
     appropriate inputs.
```
Comment (danasebo) [inline tests/matrix_a_star/test_matrix_a_star.py:324]:
```
    Dependencies:
        numpy as np
```
on L326 need indentation update

File: scalelink/main.py (hunk 5 above)
Comment (danasebo) [reply scalelink/main.py:17]: one more thing - type hinting should be removed from docstrings in this function

File: scalelink/utils/utils.py [hunk 13]
```diff
@@ -450,49 +458,50 @@ def get_input_variables(config_path):
...
-            The configs from the config file.
+      config_path (str):
+        The filepath for the config file.
 
     Dependencies:
-        configparser as cp
+      configparser as cp
+
+    Returns:
+      configs:
```
Comment (danasebo) [inline scalelink/utils/utils.py:503]: the function returns `config`, but it is `configs` it docstring

File: scalelink/indicator_matrix/indicator_matrix.py [hunk 14]
```diff
@@ -431,27 +444,27 @@ def get_deltas(df_cartesian_join, input_variables):
...
+        coefficent will be calculated for.
+      col:
+        The name of the first string column to be compared.
 
     Dependencies:
-        The feature transformer NGram from pyspark.ml.feature.
+      from pyspark.ml.feature import NGram
+
+    Returns:
+      df_compared:
```
Comment (danasebo) [inline scalelink/indicator_matrix/indicator_matrix.py]: returned variable is actually `df_with_bigrams`

**Review by danasebo — CHANGES_REQUESTED** [review]
Great work - the functions look tidy and neat!
I have a few minor suggestions to align with the agreed standards - I'm happy to approve this PR once those are addressed.

File: scalelink/indicator_matrix/indicator_matrix.py (hunk 1 above)
Comment (mary-cleaton) [reply scalelink/indicator_matrix/indicator_matrix.py]: Done.

File: tests/match_scores/test_match_scores.py (hunk 2 above)
Comment (mary-cleaton) [reply tests/match_scores/test_match_scores.py:59]: Done.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 3 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py:36]: Done.

File: scalelink/utils/utils.py (hunk 4 above)
Comment (mary-cleaton) [reply scalelink/utils/utils.py]: You're quite right! Done.

File: scalelink/main.py (hunk 5 above)
Comment (mary-cleaton) [reply scalelink/main.py:17]: Argh, legacy dependency! It used `shutil` before but now uses `boto3` and `raz_client`. Fixed.

Comment (mary-cleaton) [reply scalelink/main.py:17]: Both points done.

File: scalelink/indicator_matrix/indicator_matrix.py (hunk 6 above)
Comment (mary-cleaton) [reply scalelink/indicator_matrix/indicator_matrix.py]: It can't be `pyspark.ml.feature.NGram` as `NGram` has to be imported (it's called directly in the function). However, have changed to `NGram from pyspark.ml.feature`, which I think is more consistent yet still clear.

File: scalelink/indicator_matrix/indicator_matrix.py (hunk 7 above)
Comment (mary-cleaton) [reply scalelink/indicator_matrix/indicator_matrix.py]: Edited in the same way.

File: scalelink/indicator_matrix/indicator_matrix.py (hunk 8 above)
Comment (mary-cleaton) [reply scalelink/indicator_matrix/indicator_matrix.py]: Edited in the same way.

File: scalelink/matrix_a_star/matrix_a_star.py (hunk 9 above)
Comment (mary-cleaton) [reply scalelink/matrix_a_star/matrix_a_star.py]: Have updated the args to refer to `Njlkm` as `njklm`. Also updated where it's called in `get_matrix_a_star` (line 235 of the same script).

The reason for this (and some of the other funky variable names) is that the casing is referring to the algebra in the Scalelink paper. Njklm is actually N<sub>jklm</sub>. Likewise, that's why `K` is K not k, etc. (because k is something else in the algebra).

Comment (mary-cleaton) [reply scalelink/matrix_a_star/matrix_a_star.py]: Actually, no, on further reflection I don't want to do that (at least now). Cause if N<sub>jklm</sub> is `njklm` but K is still `K` then it's confusing. I've done a compromise (?) of renaming the arg `df_Njklm` where it's defined (line 350), used (lines 380, 382) and called (line 235), plus in the unit test (line 205 of `scalelink / tests / maktrix_a_star / test_matrix_a_star.py`).

If you feel strongly about these variables having non-standard casing, please raise an Issue so both N<sub>jklm</sub> and K can be brought into line, so everything's consistent. I'm happy to be swayed, just want it all one way or all the other.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 10 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py:150]: This looks correct to me.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 11 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py:215]: Ditto - looks correct on my end.

File: tests/matrix_a_star/test_matrix_a_star.py (hunk 12 above)
Comment (mary-cleaton) [reply tests/matrix_a_star/test_matrix_a_star.py:324]: Ditto - looks correct on my end.

File: scalelink/matrix_a_star/matrix_a_star.py (hunk 9 above)
Comment (danasebo) [reply scalelink/matrix_a_star/matrix_a_star.py]: My suggestion was to use whatever you find appropriate consistently in a docstring and a function - I agree it makes sense that these variables use capitals letters as they follow the algebra in the paper so it is an understandable exception to the rule / coding convention. Both `df_Njklm` and `Njklm` are acceptable, just choose one and use it consistently.

File: scalelink/utils/utils.py (hunk 13 above)
Comment (mary-cleaton) [reply scalelink/utils/utils.py:503]: Made consistent.

File: scalelink/indicator_matrix/indicator_matrix.py (hunk 14 above)
Comment (mary-cleaton) [reply scalelink/indicator_matrix/indicator_matrix.py]: Corrected.

**Comment by mary-cleaton** [comment]
Have addressed all comments. Now ready for @danasebo to review again.

File: scalelink/main.py (hunk 5 above)
Comment (danasebo) [reply scalelink/main.py:17]: thanks! looking at your updated dependencies, are you sure it should include `pyspark`, not `pyspark.sql.DataFrame`?

File: scalelink/matrix_a_star/matrix_a_star.py (hunk 9 above)
Comment (danasebo) [reply scalelink/matrix_a_star/matrix_a_star.py]: `calculate_njklm_values` uses `df_njklm`, whereas this function (`make_matrix_a`) uses `df_Njklm` - are you happy with this inconsistency?

File: scalelink/main.py (hunk 5 above)
Comment (mary-cleaton) [reply scalelink/main.py:17]: Fixed.

**Comment by mary-cleaton** [comment]
@danasebo - ready for you again.

File: scalelink/matrix_a_star/matrix_a_star.py (hunk 9 above)
Comment (mary-cleaton) [reply scalelink/matrix_a_star/matrix_a_star.py]: Gah! How did I not spot that before!?

Fixed by altering `calculate_njklm_values` to `df_Njklm`.

Comment (danasebo) [reply scalelink/matrix_a_star/matrix_a_star.py]: The very last thing - there are two instances of `df_njklm` in `get_matrix_a_star` function - please consider updating them for consistency. Then I think we're done here.

Comment (mary-cleaton) [reply scalelink/matrix_a_star/matrix_a_star.py]: Done!

**Review by danasebo — APPROVED** [review]
I'm happy with the additional changes - many thanks for incorporating my suggestions regarding the spotted inconsistencies.
Just to note, as expected, the main script as well as all unit tests run without any issues (no functional changes were introduced in this PR, but they were checked nevertheless, as some variable names had been updated).
